storage/innobase/btr/btr0cur.cc

```diff
@@ -3261,9 +3264,18 @@ btr_cur_ins_lock_and_undo(
 				index, thr, mtr, &prdt);
 			*inherit = false;
 		} else {
+#ifdef WITH_WSREP
+		if (!index->is_clust() && (index->type & DICT_UNIQUE) &&
```

Comment:
Why this condition needs to be this complex. If I correctly understand in case of applier and replay transaction is already committed so why we would need to take GAP-locks at all ? Here in my understanding we need to only consider conflicts between applier/replay writeset and local transaction writeset. So if applier/replay does INSERT INTO t1 VALUES (3),(5),(7) for a unique key and table already contains values (1) and (9) and local transaction tries to insert those gaps like INSERT INTO t1 values (2),(4),(8); there is no real record lock conflict but for repeatable read isolation level can we allow new values to be introduced to a range ]2,7[ ?

Why we can't just do if (wsrep_thd_is_BF()) ... ?

Reply:
The purpose of this condition is to restrict this change to affect only unique secondary index operations. Gap locks on unique secondary keys were the only problem source in multi-master high intensity conflict testing.